crates/ruff/src/rules/flake8_bandit/rules/shell_injection.rs

```diff
@@ -0,0 +1,380 @@
+                } else {
+                    Some(HasShell::Truthy)
+                }
+            }
+            ExprKind::Tuple { elts, .. } => {
+                if elts.is_empty() {
+                    Some(HasShell::Falsey)
+                } else {
+                    Some(HasShell::Truthy)
+                }
+            }
+            _ => Some(HasShell::Unknown),
+        }
+    } else {
+        None
+    }
+}
+
+fn find_shell_keyword(keywords: &[Keyword]) -> Option<&Keyword> {
+    keywords.iter().find(|keyword| {
+        keyword
+            .node
+            .arg
+            .as_ref()
+            .and_then(|arg| if arg == "shell" { Some(()) } else { None })
+            .is_some()
+    })
+}
+
+fn shell_call_looks_safe(arg: &Expr) -> bool {
+    matches!(
+        arg.node,
+        ExprKind::Constant {
+            value: Constant::Str(_),
+            ..
+        }
+    )
+}
+
+fn get_call_kind(checker: &mut Checker, func: &Expr) -> Option<CallKind> {
+    checker.ctx.resolve_call_path(func).and_then(|call_path| {
+        if CONFIG
+            .subprocess
+            .iter()
+            .any(|subprocess| call_path.as_slice() == subprocess.as_slice())
+        {
+            Some(CallKind::Subprocess)
+        } else if CONFIG
+            .shell
+            .iter()
+            .any(|shell| call_path.as_slice() == shell.as_slice())
+        {
+            Some(CallKind::Shell)
+        } else if CONFIG
+            .no_shell
+            .iter()
+            .any(|no_shell| call_path.as_slice() == no_shell.as_slice())
+        {
+            Some(CallKind::NoShell)
+        } else {
+            None
+        }
+    })
+}
+
+fn string_literal_including_list(expr: &Expr) -> Option<&str> {
+    match &expr.node {
+        ExprKind::List { elts, .. } => {
+            if elts.is_empty() {
+                None
+            } else {
+                string_literal(&elts[0])
+            }
+        }
+        _ => string_literal(expr),
+    }
+}
+
+/// S602, S603, S604, S605, S606, S607
+pub fn shell_injection(checker: &mut Checker, func: &Expr, args: &[Expr], keywords: &[Keyword]) {
+    let call_kind = get_call_kind(checker, func);
+
+    if let Some(CallKind::Subprocess) = call_kind {
+        if !args.is_empty() {
+            match has_shell(keywords) {
+                // S602
+                Some(HasShell::Truthy) => {
+                    if checker
+                        .settings
+                        .rules
+                        .enabled(Rule::SubprocessPopenWithShellEqualsTrue)
+                    {
+                        checker.diagnostics.push(Diagnostic::new(
+                            SubprocessPopenWithShellEqualsTrue {
+                                looks_safe: shell_call_looks_safe(&args[0]),
+                            },
+                            Range::from(find_shell_keyword(keywords).unwrap()),
+                        ));
+                    }
+                }
```

Comment:
It would be nice if we could avoid the `unwrap` on `find_shell_keyword`. While this is safe now, it may lead to future errors if someone changes `hash_shell`. 

How about we introduce a new `ShellKeyword` struct that has two fields: `keyword` and `has_shell` and return it from `find_shell_keyword`? We could then rewrite this to 

```
if let Some(shell_keyword) = find_shell_keyword(keywords) {
	match shell_keyword.has_shell {
		...
		
		Range::from(shell_keyword.keyword)
	}
}
```

Reply:
Good call, thanks. After applying the change I realise the `has_shell` name no longer makes sense so I've rejigged that a bit.